include/circt/Support/PrettyPrinterHelpers.h

```diff
@@ -166,6 +166,46 @@ class TokenStringSaver : public PrettyPrinter::Listener {
   void clear() override;
 };
 
+template <typename Callable, typename Data>
+class CallbackSaver {
+
+  /// List of all the unique data associated with each callback token.
+  SmallVector<Data *> list;
+  /// The storage for the callback, as a function object.
+  std::function<void(unsigned)> callbackObj;
+
+  /// Note: Callable class must implement a callable with signature::
+  /// void (Data*)
+public:
+  CallbackSaver() = default;
+  CallbackSaver(Callable &c) { init(c); }
+
+  /// Create and initialize a function object, capturing the callable object and
+  /// a poiner to the data vector. The callable object contains all the values
+  /// shared across the callback tokens. The Data list indexed by the parameter
+  /// provides the unique data for each clalback.
+  void init(Callable &c) {
```

Comment:
re:defensive, maybe assert the data list is empty?

On that, if this is meant to support changing the callback object, should there be a mechanism to reset the list?

May be a bit of a stretch, but depending on density of these callbacks, on some very large designs 32bits may not be enough?